### PR TITLE
add internal knative proxy to gloo, fix bug with routes 

### PR DESCRIPTION
**Pull request by ilackarms**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/979
resolves https://github.com/solo-io/gloo/issues/980
resolves https://github.com/solo-io/gloo/issues/981
resolves https://github.com/solo-io/gloo/issues/986
resolves https://github.com/solo-io/gloo/issues/987

### CODE REVIEW

File: projects/knative/pkg/translator/translate_test.go [hunk 1]
```diff
@@ -263,68 +262,6 @@ var _ = Describe("Translate", func() {
 										"zah.net:80",
 									},
 									Routes: []*gloov1.Route{
-										&gloov1.Route{
```
Comment (yuval-k) [inline projects/knative/pkg/translator/translate_test.go]: why is this removed?

File: projects/knative/pkg/translator/translator_syncer.go [hunk 2]
```diff
@@ -73,10 +106,21 @@ func (s *translatorSyncer) Sync(ctx context.Context, snap *v1.TranslatorSnapshot
 		return err
 	}
 
-	if err := s.propagateProxyStatus(ctx, proxy, snap.Ingresses); err != nil {
-		return errors.Wrapf(err, "failed to propagate proxy status "+
-			"to ingress objects")
-	}
+	g := &errgroup.Group{}
+	g.Go(func() error {
+		if err := s.propagateProxyStatus(ctx, externalProxy, snap.Ingresses); err != nil {
```
Comment (yuval-k) [inline projects/knative/pkg/translator/translator_syncer.go]: is nil check needed here? as externalProxy can be nil?
can this be tested?

File: projects/gloo/api/v1/settings.proto [hunk 3]
```diff
@@ -138,10 +138,15 @@ message Settings {
         // Use if running Knative Version 0.7.X or less
         string cluster_ingress_proxy_address = 1;
 
-        // Address of the knative proxy.
-        // If empty, it will default to knative-proxy.$POD_NAMESPACE.svc.cluster.local.
+        // Address of the externally-facing knative proxy.
+        // If empty, it will default to knative-external-proxy.$POD_NAMESPACE.svc.cluster.local.
         // Use if running Knative Version 0.8.X or higher
-        string knative_proxy_address = 2;
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/gloo/api/v1/settings.proto]: since knative 0.8.0 is not released yet, i dont think this is a breaking change for anyone. however i can mark it in the changelog as such if preferred

File: projects/clusteringress/pkg/translator/translate_test.go [hunk 4]
```diff
@@ -285,88 +285,6 @@ var _ = Describe("Translate", func() {
 										"zah.net:80",
 									},
 									Routes: []*gloov1.Route{
-										&gloov1.Route{
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/clusteringress/pkg/translator/translate_test.go]: the expected output was incorrect before; the scope of the route variable was one level too high (bug fix)

File: projects/clusteringress/pkg/translator/translator_syncer.go [hunk 5]
```diff
@@ -123,9 +123,10 @@ func (s *translatorSyncer) markClusterIngressesReady(ctx context.Context, cluste
 		}
 		ci.Status.InitializeConditions()
 		ci.Status.MarkNetworkConfigured()
-		ci.Status.MarkLoadBalancerReady([]knativev1alpha1.LoadBalancerIngressStatus{
+		lb := []knativev1alpha1.LoadBalancerIngressStatus{
 			{DomainInternal: s.proxyAddress},
-		})
+		}
+		ci.Status.MarkLoadBalancerReady(lb, lb, lb)
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/clusteringress/pkg/translator/translator_syncer.go]: knative wants:
param 1: the original (deprecated) load balancer status (required for older versions of knative)
param 2: the external load balancer 
param 3: the internal load balancer (i've added support for internal for 0.8.0, but older versions don't currently have it - so in this case reusing the external as the internal)

i can add support for internal load balancing in a follow-up if that's our intention

File: projects/knative/pkg/translator/translate_test.go (hunk 1 above)
Comment (ilackarms) [reply projects/knative/pkg/translator/translate_test.go]: see above comment on routes bug

File: projects/ingress/pkg/setup/setup_syncer.go [hunk 6]
```diff
@@ -39,10 +38,13 @@ import (
 	"github.com/solo-io/solo-kit/pkg/errors"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
+	knativeclientset "knative.dev/serving/pkg/client/clientset/versioned"
 )
 
 const defaultClusterIngressProxyAddress = "clusteringress-proxy." + gloodefaults.GlooSystem + ".svc.cluster.local"
-const defaultKnativeProxyAddress = "knative-proxy." + gloodefaults.GlooSystem + ".svc.cluster.local"
+
+const defaultKnativeExternalProxyAddress = "knative-external-proxy." + gloodefaults.GlooSystem + ".svc.cluster.local"
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/ingress/pkg/setup/setup_syncer.go]: 👍  will use `network.GetClusterDomainName()`

File: projects/knative/pkg/translator/translator_syncer.go (hunk 2 above)
Comment (ilackarms) [reply projects/knative/pkg/translator/translator_syncer.go]: proxy is checked for nil in `propagateProxyStatus`

File: projects/gloo/api/v1/settings.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuval-k) [reply projects/gloo/api/v1/settings.proto]: ahhh got you; this this was added in the context of knative 0.8, and not relevant to <= 0.7?
then i'm ok with this not breaking.
